src/ert/run_models/everest_run_model.py

```diff
@@ -599,3 +463,306 @@ def result(self) -> Optional[OptimalResult]:
+                assert metadata.config.variables.names is not None
+                for control_name, control_value in zip(
+                    metadata.config.variables.names,
+                    control_values[sim_idx, :],
+                    strict=False,
+                ):
+                    self._add_control(controls, control_name, control_value)
+                case_data.append((real_id, controls))
+        return case_data, active, cached
+
+    def _setup_sim(
+        self,
+        sim_id: int,
+        controls: Dict[str, Dict[str, Any]],
+        ensemble: Ensemble,
+    ) -> None:
+        def _check_suffix(
+            ext_config: "ExtParamConfig",
+            key: str,
+            assignment: Union[Dict[str, Any], Tuple[str, str], str, int],
+        ) -> None:
+            if key not in ext_config:
+                raise KeyError(f"No such key: {key}")
+            if isinstance(assignment, dict):  # handle suffixes
+                suffixes = ext_config[key]
+                if len(assignment) != len(suffixes):
+                    missingsuffixes = set(suffixes).difference(set(assignment.keys()))
+                    raise KeyError(
+                        f"Key {key} is missing values for "
+                        f"these suffixes: {missingsuffixes}"
+                    )
+                for suffix in assignment:
+                    if suffix not in suffixes:
+                        raise KeyError(
+                            f"Key {key} has suffixes {suffixes}. "
+                            f"Can't find the requested suffix {suffix}"
+                        )
+            else:
+                suffixes = ext_config[key]
+                if suffixes:
+                    raise KeyError(
+                        f"Key {key} has suffixes, a suffix must be specified"
+                    )
+
+        if set(controls.keys()) != set(self.everest_config.control_names):
+            err_msg = "Mismatch between initialized and provided control names."
+            raise KeyError(err_msg)
+
+        for control_name, control in controls.items():
+            ext_config = self.ert_config.ensemble_config.parameter_configs[control_name]
+            if isinstance(ext_config, ExtParamConfig):
+                if len(ext_config) != len(control.keys()):
+                    raise KeyError(
+                        (
+                            f"Expected {len(ext_config)} variables for "
+                            f"control {control_name}, "
+                            f"received {len(control.keys())}."
+                        )
+                    )
+                for var_name, var_setting in control.items():
+                    _check_suffix(ext_config, var_name, var_setting)
+
+                ensemble.save_parameters(
+                    control_name, sim_id, ExtParamConfig.to_dataset(control)
+                )
+
+    def create_forward_model_evaluator_function(self) -> Evaluator:
```

Comment:
Why does this need to be a factory function? `run_forward_model` can just be a method. That will also make it easier to fix the `test_simulator_cache` test.

Reply:
There is no particular reason I just kept the old structure I thought it was useful for some tests if it can be made a simple method that would be fantastic.